wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsnext/WorkspaceNextObjectsRetriever.java

```diff
@@ -0,0 +1,403 @@
+        "Service version in feature and Service objects didn't match. Service object seems broken.");
+    requireNotNullNorEmpty(
+        service.getSpec().getContainers(),
+        SERVICE_OBJECT_ERROR,
+        name,
+        version,
+        "Containers are missing.");
+    for (Container container : service.getSpec().getContainers()) {
+      requireNotNull(container, SERVICE_OBJECT_ERROR, name, version, "A container is missing.");
+      requireNotNullNorEmpty(
+          container.getImage(), SERVICE_OBJECT_ERROR, name, version, "Container image is missing.");
+    }
+  }
+
+  @VisibleForTesting
+  protected <T> T getBody(URI uri, Class<T> clas) throws IOException {
+    HttpURLConnection httpURLConnection = null;
+    try {
+      httpURLConnection = (HttpURLConnection) uri.toURL().openConnection();
+
+      int responseCode = httpURLConnection.getResponseCode();
+      if (responseCode != 200) {
+        throw new IOException(
+            format(
+                "Can't get object by URI '%s'. Error: %s",
+                uri.toString(), getError(httpURLConnection)));
+      }
+
+      return parseResponseStreamAndClose(httpURLConnection.getInputStream(), clas);
+    } finally {
+      if (httpURLConnection != null) {
+        httpURLConnection.disconnect();
+      }
+    }
+  }
+
+  private String getError(HttpURLConnection httpURLConnection) throws IOException {
+    try (InputStreamReader isr = new InputStreamReader(httpURLConnection.getInputStream())) {
+      return CharStreams.toString(isr);
+    }
+  }
+
+  protected <T> T parseResponseStreamAndClose(InputStream inputStream, Class<T> clazz)
+      throws IOException {
+    try (InputStreamReader reader = new InputStreamReader(inputStream)) {
+      T objectFromJson = GSON.fromJson(reader, clazz);
+      if (objectFromJson == null) {
+        throw new IOException(
+            "Internal server error. Unexpected response body received from Feature API.");
+      }
+      return objectFromJson;
+    } catch (JsonParseException e) {
+      throw new IOException(e.getLocalizedMessage(), e);
+    }
+  }
+
+  private void requireNotNullNorEmpty(String s, String error, String... errorArgs)
+      throws InfrastructureException {
+    if (s == null || s.isEmpty()) {
+      throw new InfrastructureException(format(error, (Object[]) errorArgs));
+    }
+  }
+
+  private void requireNotNullNorEmpty(Collection objects, String error, String... errorArgs)
+      throws InfrastructureException {
+    if (objects == null || objects.isEmpty()) {
+      throw new InfrastructureException(format(error, (Object[]) errorArgs));
+    }
+  }
+
+  private void requireNotNull(Object object, String error, String... errorArgs)
+      throws InfrastructureException {
+    if (object == null) {
+      throw new InfrastructureException(format(error, (Object[]) errorArgs));
+    }
+  }
+
+  private void requireEqual(String version, String version1, String error, String... errorArgs)
```

Comment:
I guess you used `Refactor` functionality and it's why the arguments are named `version` and `version1`. Am I right?

Reply:
Looks like, I don't really remember, to be honest.